pkg/api/file.go

```diff
@@ -198,6 +200,59 @@ func (s *server) fileUploadHandler(w http.ResponseWriter, r *http.Request) {
 	})
 }
 
+// fileUploadInfo contains the data for a file to be uploaded
+type fileUploadInfo struct {
+	name        string // file name
+	size        int64  // file size
+	contentType string
+	toEncrypt   bool
+	reader      io.Reader
+}
+
+// storeFile uploads the given file and returns its reference
+// this function was extracted from `fileUploadHandler` and should eventually replace its current code
+func storeFile(ctx context.Context, fileInfo *fileUploadInfo, s storage.Storer) (swarm.Address, error) {
+	// first store the file and get its reference
+	sp := splitter.NewSimpleSplitter(s)
+	fr, err := file.SplitWriteAll(ctx, sp, fileInfo.reader, fileInfo.size, fileInfo.toEncrypt)
```

Comment:
the `toEncrypt` here can be removed if it is set on the context

Reply:
if i remove `toEncrypt` from `fileInfo`, why not also `size` and `reader`?

i think this discussion is healthy to have, but outside of the scope for this PR for now. we have follow-up issues where we can work on this stuff.

Reply:
so the reasoning ass follows:
context passing around of values is meant to pass around _request specific_ state and can pass key/value pairs (it is not a map tho); a value _might_ be found on a context, there is not a way to assure that a certain key/value pair will be on the context object except for verifying and tightly guarding the points from which the context propagates that the values are indeed there, propagating correctly to the entire cascade of calls. this is quite difficult to maintain as an approach. theoretically, you can create a context that will carry around even pointers to certain data structs (or readers in this case). but as i said there is no way of verifying that they will be there, thus making the code in the cascade of calls clunky will all sorts of possible `nil` checks and typecasts that slow down execution, etc. the `toEncrypt` value though, is a classic example of what _can_ be passed on context, since it is an optional parameter and you can assume it is false as long as it is not declaratively specified otherwise. `size` and `reader` do not have the same status IMO and trying to pass them around is not really idiomatic go (i've never seen anything like that anywhere), however certainly doable.

Reply:
also:
https://medium.com/@matryer/context-keys-in-go-5312346a868d
https://stackoverflow.com/questions/40379960/golang-context-withvalue-how-to-add-several-key-value-pairs

Reply:
i'm convinced. i've removed this field from the file info struct and am extracting it from the context now. 

thanks for the links too